src/program/posix/ast.rs

```diff
@@ -0,0 +1,140 @@
+//! Abstract Syntax Tree for the POSIX language.
+use program::ast::Interpreter;
+
+/// A program is the result of parsing a sequence of commands.
+#[derive(Debug, Clone)]
+pub struct Program(pub Vec<Box<Command>>);
+
+/// A program's text and the interpreter to be used.
+// TODO #8: Include grammar separate from interpreter?
+#[derive(Debug, Clone)]
+pub struct BridgedProgram(pub Interpreter, pub String);
+
+/// A command is a *highly* mutually-recursive node with the main features
+/// of the POSIX language.
+#[derive(Debug, Clone)]
+pub enum Command {
+    /// Just a single command, with it's arguments.
+    ///
+    /// ### Examples
+    ///
+    /// ```sh
+    /// date --iso-8601
+    /// ```
+    // TODO #8: Simple should not just be a vec of words.
+    Simple(Vec<Word>),
+
+    /// A full program embedded in a compound command.
+    ///
+    /// ```sh
+    /// { ls ; }
+    /// ```
+    Compound(Vec<Box<Command>>),
+
+    /// Performs boolean negation to the status code of the inner
+    /// command.
+    ///
+    /// ### Examples
+    ///
+    /// ```sh
+    /// ! grep 'password' data.txt
+    /// ```
+    Not(Box<Command>),
+
+    /// Perform the first command, conditionally running the next
+    /// upon success.
+    ///
+    /// ### Examples
+    ///
+    /// ```sh
+    /// mkdir tmp && cd tmp
+    /// ```
+    And(Box<Command>, Box<Command>),
+
+    /// Perform the first command, conditionally running the next
+    /// upon failure.
+    ///
+    /// ### Examples
+    ///
+    /// ```sh
+    /// kill $1 || kill -9 $1
+    /// ```
+    Or(Box<Command>, Box<Command>),
+
+    Cond(Box<Command>, Box<Command>, Option<Box<Command>>),
```

Comment:
Should be able to avoid this by constructing either `And` or `Or` nodes, or a combination of the two.